### PR TITLE
Feat ledger

### PR DESCRIPTION
**Pull request by DZahar0v**

### CODE REVIEW

File: contracts/Ledger.sol [hunk 1]
```diff
@@ -232,91 +231,69 @@ contract Ledger {
...
+            // NOTE: if ledger stake > active balance we are trying to bond all funds
+            uint128 diff = _ledgerStake - activeBalance;
+            uint128 diffToRebond = diff > unlockingBalance ? unlockingBalance : diff;
+            if (diffToRebond > 0) {
+                CONTROLLER.rebond(diffToRebond, MAX_UNLOCKING_CHUNKS);
+                diff -= diffToRebond;
             }
 
-            uint128 relayFreeBalance = _report.getFreeBalance();
             if ((relayFreeBalance == transferUpwardBalance) && (transferUpwardBalance > 0)) {
```
Comment (georgiypetrov) [inline contracts/Ledger.sol:254]: comment - massage

File: contracts/Ledger.sol [hunk 2]
```diff
@@ -232,91 +231,69 @@ contract Ledger {
...
                 // In case if bond amount = transferUpwardBalance we can't distinguish 2 messages were success or 2 masseges were failed
```
Comment (georgiypetrov) [inline contracts/Ledger.sol:255]: massege

File: contracts/Ledger.sol [hunk 3]
```diff
@@ -232,91 +231,69 @@ contract Ledger {
...
-                pendingBonds = relayFreeBalance;
-            } else if (_report.stakeStatus == Types.LedgerStatus.None && relayFreeBalance >= MIN_NOMINATOR_BALANCE) {
-                CONTROLLER.bond(controllerAccount, relayFreeBalance);
-                pendingBonds = relayFreeBalance;
+            if ((diff > 0) && (relayFreeBalance > 0)) {    
+                uint128 diffToBond = diff > relayFreeBalance ? relayFreeBalance : diff;
+                if (_report.stakeStatus == Types.LedgerStatus.Nominator || _report.stakeStatus == Types.LedgerStatus.Idle) {
+                    CONTROLLER.bondExtra(diffToBond);
+                    pendingBonds = diffToBond;
+                } else if (_report.stakeStatus == Types.LedgerStatus.None && relayFreeBalance >= MIN_NOMINATOR_BALANCE) {
```
Comment (georgiypetrov) [inline contracts/Ledger.sol:264]: diffToBond?

Comment (DZahar0v) [reply contracts/Ledger.sol:264]: Can you please clarify your question?

**Review by karapetyantv — COMMENTED** [review]
https://github.com/mixbytes/lido-dot-ksm/commit/6219ee363a60b6f8fc12b1a67aa528d0203d5b09 reverts changes made in https://github.com/mixbytes/lido-dot-ksm/commit/949b54ae88ba3825a9a783c9af3ebb8d33225b34. Is it intended?

File: contracts/Ledger.sol (hunk 2 above)
Comment (DZahar0v) [reply contracts/Ledger.sol:255]: Done

File: contracts/Ledger.sol (hunk 3 above)
Comment (DZahar0v) [reply contracts/Ledger.sol:264]: Done

**Comment by algys** [comment]
> 949b54a

[949b54a](https://github.com/mixbytes/lido-dot-ksm/pull/34/commits/949b54ae88ba3825a9a783c9af3ebb8d33225b34)

absolutely agree, seems like branch improperly rebased, at least [949b54a](https://github.com/mixbytes/lido-dot-ksm/pull/34/commits/949b54ae88ba3825a9a783c9af3ebb8d33225b34) this commit contain unresolved conflict

try to remove this commit from history

File: contracts/Ledger.sol [hunk 4]
```diff
@@ -232,91 +231,69 @@ contract Ledger {
...
+            if (diff > 0) {
+                CONTROLLER.unbond(diff);
             }
 
-            // withdraw if we have some unlocked
-            if (deficit > 0 && withdrawableBalance > 0) {
+            // NOTE: if ledger stake == active balance we only withdraw unlocked balance
+            if (withdrawableBalance > 0) {
                 uint32 slashSpans = 0;
                 if ((_report.unlocking.length == 0) && (_report.activeBalance <= MINIMUM_BALANCE)) {
```
Comment (algys) [inline contracts/Ledger.sol:285]: excess brakets

File: contracts/Ledger.sol [hunk 5]
```diff
@@ -232,91 +231,69 @@ contract Ledger {
...
-        if (_report.stashBalance <= _ledgerStake) {
-            //    Staking strategy:
-            //     - upward transfer deficit tokens
-            //     - rebond all unlocking tokens
-            //     - bond_extra all free balance
-
+        // Always transfer deficit to relay chain
+        if (_report.stashBalance < _ledgerStake) {
             uint128 deficit = _ledgerStake - _report.stashBalance;
+            LIDO.transferToLedger(deficit);
```
[the remark this answers is not in the document]
Comment (DZahar0v) [reply contracts/Ledger.sol:237]: Done

File: contracts/Ledger.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (DZahar0v) [reply contracts/Ledger.sol:254]: Done

File: contracts/Ledger.sol (hunk 4 above)
Comment (DZahar0v) [reply contracts/Ledger.sol:285]: Done

**Review by algys — APPROVED** [review]

**Pull request merged**